lib/puppet/pops/types/iterable.rb

```diff
@@ -202,6 +202,26 @@ def method_missing(name, *arguments, &block)
       @enumeration.send(name, *arguments, &block)
     end
 
+    def next
+      @enumeration.next
+    end
+
+    def map(*args, &block)
+      @enumeration.map(*args, &block)
+    end
+
+    def reduce(*args, &block)
+      @enumeration.reduce(*args, &block)
+    end
+
+    def all?(&block)
+      @enumeration.all?(&block)
+    end
+
+    def any?(&block)
+      @enumeration.any?(&block)
+    end
+
```

Comment:
What are these for?

Reply:
We define Puppet functions like `reduce` that end up calling reduce on the iterable they have. However, Iterator doesn't define reduce (or next, or map, etc) it uses `method_missing` to forward calls that aren't defined on it to its wrapped enumeration. Since we know we call these methods in core Puppet, instead of letting them percolate up to method_missing this just directly forwards the to the enumeration.

Reply:
gotcha, thanks! I haven't had the (mis)fortune of getting to know `method_missing` very intimately